crates/hir-def/src/lib.rs

```diff
@@ -156,19 +156,24 @@ impl<N: ItemTreeNode> Hash for ItemLoc<N> {
     }
 }
 
-#[derive(Debug)]
-pub struct AssocItemLoc<N: ItemTreeNode> {
-    pub container: ItemContainerId,
-    pub id: ItemTreeId<N>,
+#[derive(Debug, Clone, Copy)]
+pub struct InheritedVisibilityLoc {
+    pub raw_visibility_id: RawVisibilityId,
+    pub tree_id: TreeId,
 }
 
-impl<N: ItemTreeNode> Clone for AssocItemLoc<N> {
-    fn clone(&self) -> Self {
-        Self { container: self.container, id: self.id }
+impl InheritedVisibilityLoc {
+    pub fn new(visibility_id: RawVisibilityId, tree_id: TreeId) -> Self {
+        Self { raw_visibility_id: visibility_id, tree_id }
     }
 }
 
-impl<N: ItemTreeNode> Copy for AssocItemLoc<N> {}
+#[derive(Debug, Clone, Copy)]
+pub struct AssocItemLoc<N: ItemTreeNode> {
+    pub container: ItemContainerId,
+    pub id: ItemTreeId<N>,
+    pub inherited_visibility: Option<InheritedVisibilityLoc>,
```

Comment:
sorry, but this is not the right approch. The `Loc` just contain the necessary data to identify the item in question; the visibility should not be a part of them. I should have looked into this a bit more before making vague comments, because as you found out the collector can't really change the visibility :sweat_smile: 

I think the right place to handle this is in the `fn_data_query`, `const_data_query` and `type_alias_data_query`, which currently take the visibility from the item tree and return it:
https://github.com/rust-lang/rust-analyzer/blob/65874dfff205335840cc9265ba06d74f234b719d/crates/hir-def/src/data.rs#L96
It should be possible to simply check whether the container is a trait and inherit the visibility if necessary there.

Reply:
Oh yes. I just made it complex. Good to learn the design by making mistakes :)